exporter/chronicleexporter/marshal.go

```diff
@@ -0,0 +1,208 @@
+	Timestamp string `json:"ts_rfc3339"`
+}
+
+func (ce *marshaler) extractRawLogs(ld plog.Logs) ([]entry, error) {
+	entries := []entry{}
+
+	for i := 0; i < ld.ResourceLogs().Len(); i++ {
+		resourceLog := ld.ResourceLogs().At(i)
+		for j := 0; j < resourceLog.ScopeLogs().Len(); j++ {
+			scopeLog := resourceLog.ScopeLogs().At(j)
+			for k := 0; k < scopeLog.LogRecords().Len(); k++ {
+				logRecord := scopeLog.LogRecords().At(k)
+
+				var rawLog string
+				var err error
+				if ce.cfg.RawLogField == "" {
+					body := logRecord.Body().Str()
+					entireLogRecord := map[string]any{
+						"body":       body,
+						"attributes": logRecord.Attributes().AsRaw(),
+					}
+
+					bytesLogRecord, err := json.Marshal(entireLogRecord)
+					if err != nil {
+						return nil, fmt.Errorf("marshal log record: %w", err)
+					}
+
+					rawLog = string(bytesLogRecord)
+				} else {
+					rawLog, err = ce.getRawField(logRecord)
+					if err != nil {
+						return nil, fmt.Errorf("get raw field: %w", err)
+					}
+				}
+
+				entries = append(entries, entry{
+					LogText:   rawLog,
+					Timestamp: logRecord.Timestamp().AsTime().Format(time.RFC3339Nano),
+				})
+			}
+		}
+	}
+
+	return entries, nil
+}
+
+func (ce *marshaler) getRawField(logRecord plog.LogRecord) (string, error) {
+	topLevelField, nestedFields, err := parseLogField(ce.cfg.RawLogField)
+	if err != nil {
+		return "", err
+	}
+
+	if len(nestedFields) == 0 {
+		return ce.getTopLevelFieldAsString(logRecord, topLevelField)
+	}
+
+	var logMap map[string]any
+	switch topLevelField {
+	case "attributes":
+		logMap = logRecord.Attributes().AsRaw()
+	case "body":
+		if logRecord.Body().Type() != pcommon.ValueTypeMap {
+			return "", errors.New("body is not a map")
+		}
+		logMap = logRecord.Body().Map().AsRaw()
+	default:
+		return "", fmt.Errorf("unsupported top level field: %s", topLevelField)
+	}
+
+	return extractNestedValue(logMap, nestedFields)
+}
+
+func (ce *marshaler) getTopLevelFieldAsString(logRecord plog.LogRecord, field string) (string, error) {
+	switch field {
+	case "attributes":
+		attributes := logRecord.Attributes().AsRaw()
+		bytes, err := json.Marshal(attributes)
+		if err != nil {
+			return "", fmt.Errorf("marshal attributes: %w", err)
+		}
+		return string(bytes), nil
+	case "body":
+		switch logRecord.Body().Type() {
+		case pcommon.ValueTypeStr:
+			return logRecord.Body().Str(), nil
+		case pcommon.ValueTypeMap:
+			bodyMap := logRecord.Body().Map().AsRaw()
+			bytes, err := json.Marshal(bodyMap)
+			if err != nil {
+				return "", fmt.Errorf("marshal body map: %w", err)
+			}
+			return string(bytes), nil
+		default:
+			return "", errors.New("unsupported body type")
+		}
+	default:
+		return "", fmt.Errorf("unsupported top level field: %s", field)
+	}
+}
+
+func parseLogField(field string) (string, []string, error) {
+	parts := strings.SplitN(field, `["`, 2)
+
+	if len(parts) == 1 {
+		return parts[0], nil, nil
+	}
+
+	re := regexp.MustCompile(`\["(.*?)"\]`)
```

Comment:
This should be hoisted out of this function so it is compiled only once.

Reply:
Updated.